### PR TITLE
[WIP][RFC] SMOTE oversampling

### PR DESCRIPTION
**Pull request by matejklemen**
##### Issue
Implements SMOTE (or rather the SMOTE-NC variation) - relevant issues: #3198, #2436.

##### Description of changes
This pull request adds the [SMOTE oversampling method](https://jair.org/index.php/jair/article/view/10302) and an interface to use it via Data sampler.
It requires more testing and likely some other changes, but I wanted to get some feedback and ask a question.

The main thing that I'm not sure about is the way output should be provided. With other methods, there's a pretty clear distinction between what belongs in the _sample_ and what in the _remaining data_ (the 2 ouputs of Data sampler).  
With oversampling, I see two ways to divide up the data:
(1.) Putting the existing data, merged with newly generated data all into the 'sample' output slot and None into the 'remaining' output slot (**current way**) or
(2.) putting the newly generated data into into the 'sample' output slot and the existing data into the 'remaining' output slot.  
I would like to get a second opinion / new suggestion on how this should be done. 

General comments are also much appreciated.

##### Things I am aware I need to fix:
- more testing (particularly the case where input is a SQLTable, which I've not given any attention yet),
- move the SMOTE logic into a separate function (i.e. outside of `__call__()`),
- include meta attributes in distance calculation (same way as categorical ones),
- figure out how to move generation of new data into a new thread (GUI currently freezes on bigger data sets). 

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by janezd** [comment]
I don't think this is a good idea.

First, I don't know how this widget would be used. I imagine that the only use of such data would be for learning. However, oversampled data must not be used in cross validation or similar schema because it would result in having duplicates of training data in test data. Oversampling would have to happen within cross validation, on training data only. But for this, the method would have to be implemented as a preprocessor, and not within this widget.

Second, I checked the paper and I don't understand how they managed to publish it. If I understood it correctly, they published ROC curves in which individual points correspond to different oversampling settings. In other words, their "ROC curves" (which are not really ROC curves in the usual meaning) show how oversampling affects cross validation -- whatever this is supposed to show. Unless I'm mistaken about this, I really wonder how could this get past reviewers.

**Comment by janezd** [comment]
We discussed this PR at today's meeting.
 
1. Oversampling does not belong to this widget. *If anywhere*, it should be in preprocessing.
2. *If anywhere* refers to the observation that oversampling does not work (except perhaps in papers about oversampling).
3. If we still decide we want to have it in Orange, we should select an appropriate algorithm. Based on my reading the paper, SMOTE's approach doesn't look OK, and it was not properly tested, either.

**Comment by janezd** [comment]
It would be if it worked. My observation is that it works just in paper about oversampling.

Balancing imbalanced data results in models that are incorrectly biased towards the minority class. With 95 % majority, it makes perfect sense to always predict the majority class ... unless your goal is detecting rare cases (which it actually probably is). But then you should not use algorithms that were designed for classification and that optimize classification accuracy (balancing will decrease it), predicted probabilities (balancing will obviously change them) or a similar score. Balancing data won't translate a problem of detecting anomalies into a classification problem.

Thus oversampling (although in such a way that will only oversample the training data) would lead to bad practice, while I don't see any scenarios in which this would actually be useful.

One could object that Orange is just a tool and it's the user's responsibility to not misuse it. I can think of two other issues in which we had similar discussions. But think about programming languages: modern languages are all about null-safety, though one could say it's user's responsibility to not dereference them. Kotlin's switch requires listing all cases (or explicitly give a default); wouldn't it be the programmers responsibility to cover all that can happen? Language designers would be hesitant to add features that are unreadable, confusing, ambiguous. I believe that some reasonable guidance is a good thing.

**Comment by AuSpotter** [comment]
> It would be if it worked. My observation is that it works just in paper about oversampling.
> 
> Balancing imbalanced data results in models that are incorrectly biased towards the minority class. With 95 % majority, it makes perfect sense to always predict the majority class ... unless your goal is detecting rare cases (which it actually probably is). But then you should not use algorithms that were designed for classification and that optimize classification accuracy (balancing will decrease it), predicted probabilities (balancing will obviously change them) or a similar score. Balancing data won't translate a problem of detecting anomalies into a classification problem.
> 
> Thus oversampling (although in such a way that will only oversample the training data) would lead to bad practice, while I don't see any scenarios in which this would actually be useful.
> 
> One could object that Orange is just a tool and it's the user's responsibility to not misuse it. I can think of two other issues in which we had similar discussions. But think about programming languages: modern languages are all about null-safety, though one could say it's user's responsibility to not dereference them. Kotlin's switch requires listing all cases (or explicitly give a default); wouldn't it be the programmers responsibility to cover all that can happen? Language designers would be hesitant to add features that are unreadable, confusing, ambiguous. I believe that some reasonable guidance is a good thing.

Hi @janezd - actually the problem of imbalanced classes is a very relevant and important problem in the field of geoscience data analysis. Addressing imbalance is fundamentally important, e.g., problems that relate to the prediction of rock types from geophysical data (as one example [https://www.sciencedirect.com/science/article/pii/S2590197419300011](url). Note the use of SMOTE in this paper

** I will freely admit ** that SMOTE isn't a perfect solution, but those of us in the geoscience/spatial data fields need (at least) a way to randomly undersample majority classes; optimally a robust way to balance the classes in a sensible way

**Comment by bellini1** [comment]
Hi. Emergency physician, pandemic advisor and Ai practitioner: We really need SMOTE for health care research and modeling of the pandemic. As above, not perfect but the vast majority of our outcome classes are undersampled and we need to figure out all sorts of synthetic data methods to improve our models and save lives, SMOTE, digital twins etc.
Please think about adding, Orange folks!
Thanks

**Comment by AuSpotter** [comment]
Wow - that's a much more important use-case than my rock related one!

**Comment by AuSpotter** [comment]
Is it possible to re-open this request, @janezd ?
